Bindings for `std::unordered_map` are wrong

```c++
#include <unordered_map>
char (*__kaboom)[sizeof(std::unordered_map<int, int>)] = 1;
```
(Trick taken from https://stackoverflow.com/a/35261673/1021344)

This will print:
```console
clang++ uno.cpp
uno.cpp:3:8: error: cannot initialize a variable of type 'char (*)[40]' with an rvalue of type 'int'
char (*__kaboom)[sizeof(std::unordered_map<int, int>)] = 1;
       ^                                                 ~
1 error generated.
```

So the size is 40. However our bindings only have [*a single pointer*](https://github.com/bpfkorea/agora/blob/eb4ab3b0083c72fa2b7eb815ad2cd7a7290ffad4/source/scpd/Cpp.d#L85-L107).

That's on OSX with the latest clang++. I expect things to be wrong on other platforms too.

--- a/.github/workflows/documentation.yml
+++ b/.github/workflows/documentation.yml
@@ -34,7 +34,7 @@ jobs:
           sudo apt-get update
           sudo apt-get install libsqlite3-dev
           sudo apt-get install libsodium-dev
-          sudo apt-get install g++-9
+          sudo apt-get install clang
 
       - name: Build documentation
         run: |

--- a/.github/workflows/integration.yml
+++ b/.github/workflows/integration.yml
@@ -25,7 +25,7 @@ jobs:
     - name: Install dependencies
       run: |
         sudo apt-get update
-        sudo apt-get install g++-9 libsodium-dev libsqlite3-dev
+        sudo apt-get install libsodium-dev libsqlite3-dev clang
 
     - name: Checkout faucet
       uses: actions/checkout@v2

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -65,7 +65,7 @@ jobs:
       if: runner.os == 'Linux'
       run: |
         sudo apt-get update
-        sudo apt-get install g++-9 libsodium-dev libsqlite3-dev
+        sudo apt-get install libsodium-dev libsqlite3-dev clang
 
     - name: '[Windows] Install dependencies & setup environment'
       if: runner.os == 'Windows'

--- a/scripts/build_tracy.d
+++ b/scripts/build_tracy.d
@@ -43,8 +43,8 @@ else
 {
     immutable ObjExt = `.o`;
     immutable CppCmd = [
-        `gcc`, `-std=c++14`, `-g`, `-c`, `-DTRACY_ENABLE`,
-        `-o`, OutputFile, TracySourcePath,
+        `clang++`, `-std=c++14`, `-g`, `-c`, `-DTRACY_ENABLE`,
+        `-fPIC`, `-o`, OutputFile, TracySourcePath,
     ];
 }
 

--- a/source/agora/consensus/Quorum.d
+++ b/source/agora/consensus/Quorum.d
@@ -611,7 +611,7 @@ private void verifyQuorumsIntersect (QuorumConfig[PublicKey] quorums)
 {
     import scpd.quorum.QuorumIntersectionChecker;
 
-    auto qm = QuorumTracker.QuorumMap.create();
+    auto qm = QuorumTracker.QuorumMap(CppCtor.Use);
     foreach (key, quorum; quorums)
     {
         auto scp = toSCPQuorumSet(quorum);

--- a/source/agora/utils/SCPPrettyPrinter.d
+++ b/source/agora/utils/SCPPrettyPrinter.d
@@ -107,7 +107,7 @@ private struct QuorumFmt
     {
         try
         {
-            SCPQuorumSetPtr qset;
+            SCPQuorumSetPtr qset = SCPQuorumSetPtr(CppCtor.Use);
             if (this.getQSet !is null)
                 qset = this.getQSet(this.hash);
 

--- a/source/scpd/Cpp.d
+++ b/source/scpd/Cpp.d
@@ -34,6 +34,76 @@ import std.meta;
 
 import vibe.data.json;
 
+public enum CppCtor { Use = 0 }
+
+extern(C++) {
+@trusted nothrow @nogc:
+    void defaultCtorCPPObject(T) (T* ptr);
+    void dtorCPPObject(T) (T* ptr);
+    void opAssignCPPObject(T) (T* lhs, T* rhs);
+    void copyCtorCPPObject(T) (T* ptr, inout(T)* rhs);
+    void copyCtorCPPObject(T) (immutable(T)* ptr, inout(T)* rhs);
+    int getCPPSizeof(T) ();
+}
+
+private mixin template CPPBindingMixin (T, bool Copyable = true, bool DefaultConstructable = false)
+{
+    extern(D)
+    {
+        static if (!DefaultConstructable) @disable this();
+        this (CppCtor use) @trusted nothrow @nogc
+        {
+            defaultCtorCPPObject!T(&this);
+        }
+
+        ~this () @trusted nothrow @nogc
+        {
+            dtorCPPObject!T(&this);
+        }
+
+        static if (Copyable)
+        this (ref return scope inout T rhs) @trusted nothrow @nogc
+        {
+            copyCtorCPPObject!T(&this, &rhs);
+        }
+
+        static if (Copyable)
+        this (ref return scope inout T rhs) immutable @trusted nothrow @nogc
+        {
+            copyCtorCPPObject!T(&this, &rhs);
+        }
+
+        static if (Copyable)
+        void opAssign()(auto ref T rhs) @trusted nothrow @nogc
+        {
+            opAssignCPPObject!T(&this, &rhs);
+        }
+    }
+}
+
+/// Can't import `core.stdcpp.allocator` because it transitively imports
+/// `core.stdcpp.exception`
+/// In this case we just need to get the name right for `vector`
+
+extern(C++, (StdNamespace)) extern(C++, class) struct allocator (T) {}
+
+extern(C++, (StdNamespace)) extern(C++, class) struct less (T) {}
+
+extern(C++, (StdNamespace)) extern(C++, class) struct default_delete (T) {}
+
+// simplistic std::pair bindings
+public extern(C++, (StdNamespace)) struct pair (T1, T2)
+{
+    T1 first;
+    T2 second;
+}
+
+// fake std::hash (for mangling)
+public extern(C++, (StdNamespace))  struct hash (T) {}
+
+// fake std::equal_to (for mangling)
+public extern(C++, (StdNamespace))  struct equal_to (T = void) {}
+
 extern(C++, (StdNamespace)) {
     /// Simple binding to `std::shared_ptr`
     extern(C++, class) struct shared_ptr (T)
@@ -43,19 +113,22 @@ extern(C++, (StdNamespace)) {
         else
             private alias TPtr = T*;
 
-        ~this() @safe {}
+        mixin CPPBindingMixin!(shared_ptr!T);
+
         TPtr ptr;
         void* _control_block;
         alias ptr this;
     }
 
     /// Simple binding to `std::unique_ptr`
-    extern(C++, class) struct unique_ptr (T)
+    extern(C++, class) struct unique_ptr (T, Deleter = default_delete!T)
     {
         static if (is(T == class) || is(T == interface))
             private alias TPtr = T;
         else
             private alias TPtr = T*;
+
+        mixin CPPBindingMixin!(unique_ptr!T, false);
 
         TPtr ptr;
         alias ptr this;
@@ -76,49 +149,43 @@ private nothrow @nogc extern(C++) void cpp_unordered_map_assign (K, V)(
 private pure nothrow @nogc @safe extern(C++) size_t cpp_unordered_map_length (K, V)(
     const(void)* map);
 
-// @bug with substitution
-// https://issues.dlang.org/show_bug.cgi?id=20679
-//private pure nothrow @nogc @safe extern(C++) unordered_map!(K, V)* cpp_unordered_map_create (K, V)();
-
-/// Create a new unordered map with the given Key/T types, and return a pointer to it
-private pure nothrow @nogc @safe extern(C++) void* cpp_unordered_map_create (K, V)();
-
 /// Rudimentary bindings for std::unordered_map
 extern(C++, (StdNamespace))
 public struct unordered_map (Key, T, Hash = hash!Key, KeyEqual = equal_to!Key, Allocator = allocator!(pair!(const Key, T)))
 {
-    private void* ptr;
+    version (CppRuntime_Clang)
+        private ulong[40 / ulong.sizeof] _data;
+    else
+        private ulong[56 / ulong.sizeof] _data;
 
-    extern(D) void opIndexAssign (in T value, in Key key) @trusted @nogc nothrow
+    mixin CPPBindingMixin!(unordered_map!(Key, T));
+
+    void opIndexAssign (in T value, in Key key) @trusted @nogc nothrow
     {
-        cpp_unordered_map_assign!(Key, T)(this.ptr, key, value);
+        cpp_unordered_map_assign!(Key, T)(&this, key, value);
     }
 
-    extern(D) size_t length () pure nothrow @safe @nogc
+    size_t length () pure nothrow @safe @nogc
     {
-        return cpp_unordered_map_length!(Key, T)(this.ptr);
-    }
-
-    /// create a new unordered map allocated on the C++ side.
-    /// Note that we currently only use this in unittests
-    extern(D) static unordered_map create () pure @trusted @nogc nothrow
-    {
-        return unordered_map(cpp_unordered_map_create!(Key, T)());
+        return cpp_unordered_map_length!(Key, T)(&this);
     }
 }
 
 unittest
 {
-    auto map = unordered_map!(int, int).create();
+    auto map = unordered_map!(int, int)(CppCtor.Use);
     assert(map.length() == 0);
     map[1] = 1;
     assert(map.length() == 1);
+    auto copy = map;
+    assert(copy.length() == map.length());
+
+    auto copy2 = unordered_map!(int, int)(CppCtor.Use);
+    copy2 = copy;
+    assert(copy2.length() == map.length());
 }
 
 extern(C++, `std`) {
-    /// Binding: Needs to be instantiated on C++ side
-    shared_ptr!T make_shared(T, Args...)(Args args);
-
     class runtime_error : exception { }
     class logic_error : exception { }
 
@@ -128,11 +195,21 @@ extern(C++, `std`) {
         this() nothrow {}
         const(char)* what() const nothrow;
     }
+}
+
+extern(C++, (StdNamespace)) {
+    /// Binding: Needs to be instantiated on C++ side
+    shared_ptr!T make_shared(T, Args...)(Args args);
 
     /// Fake bindings for std::set
-    public extern(C++, class) struct set (Key)
+    public extern(C++, class) struct set (Key, Compare = less!Key, Allocator = allocator!Key)
     {
-        void*[3] ptr;
+        version (CppRuntime_Clang)
+            private ulong[24 / ulong.sizeof] _data;
+        else
+            private ulong[48 / ulong.sizeof] _data;
+
+        mixin CPPBindingMixin!(set!Key);
 
         /// Foreach support
         extern(D) public int opApply (scope int delegate(ref const(Key)) dg) const
@@ -152,12 +229,17 @@ extern(C++, `std`) {
         {
             return cpp_set_empty!Key(cast(const void*)&this);
         }
-    }
+}
 
     /// Fake bindings for std::map
-    public extern(C++, class) struct map (Key, Value)
+    public extern(C++, class) struct map (Key, Value, Compare = less!Key, Allocator = allocator!(pair!(const Key, Value)))
     {
-        void*[3] ptr;
+        version (CppRuntime_Clang)
+            private ulong[24 / ulong.sizeof] _data;
+        else
+            private ulong[48 / ulong.sizeof] _data;
+
+        mixin CPPBindingMixin!(map!(Key, Value));
     }
 
     // only used at compile-time on the C++ side, here for mangling
@@ -209,25 +291,6 @@ version (Posix)
         assert(values == [1, 2, 3, 4, 5]);
     }
 }
-
-/// Can't import `core.stdcpp.allocator` because it transitively imports
-/// `core.stdcpp.exception`
-/// In this case we just need to get the name right for `vector`
-
-extern(C++, (StdNamespace)) extern(C++, class) struct allocator (T) {}
-
-// simplistic std::pair bindings
-public extern(C++, (StdNamespace)) struct pair (T1, T2)
-{
-    T1 first;
-    T2 second;
-}
-
-// fake std::hash (for mangling)
-public extern(C++, (StdNamespace))  struct hash (T) {}
-
-// fake std::equal_to (for mangling)
-public extern(C++, (StdNamespace))  struct equal_to (T = void) {}
 
 /*******************************************************************************
 
@@ -468,3 +531,11 @@ unittest
 
 /// Invoke an std::function pointer (note: must be void* due to mangling issues)
 extern(C++) void callCPPDelegate (void* cb);
+
+unittest
+{
+    assert(unordered_map!(int,int).sizeof == getCPPSizeof!(unordered_map!(int,int))());
+    assert(shared_ptr!int.sizeof == getCPPSizeof!(shared_ptr!int)());
+    assert(set!int.sizeof == getCPPSizeof!(set!int)());
+    assert(map!(int,int).sizeof == getCPPSizeof!(map!(int,int))());
+}

--- a/source/scpd/quorum/QuorumTracker.d
+++ b/source/scpd/quorum/QuorumTracker.d
@@ -71,5 +71,8 @@ public:
     ref const(QuorumMap) getQuorum () const;
 }
 
-// should be 48, see #753
-static assert(QuorumTracker.sizeof == 16);
+extern (D):
+unittest
+{
+    assert(QuorumTracker.sizeof == getCPPSizeof!QuorumTracker());
+}

--- a/source/scpd/scp/BallotProtocol.d
+++ b/source/scpd/scp/BallotProtocol.d
@@ -299,4 +299,8 @@ extern(C++, class) public struct BallotProtocol
     void checkHeardFromQuorum();
 }
 
-static assert(BallotProtocol.sizeof == 136);
+extern (D):
+unittest
+{
+    assert(BallotProtocol.sizeof == getCPPSizeof!BallotProtocol());
+}

--- a/source/scpd/scp/NominationProtocol.d
+++ b/source/scpd/scp/NominationProtocol.d
@@ -118,4 +118,8 @@ nothrow:
     vector!SCPEnvelope getCurrentState() const;
 }
 
-static assert(NominationProtocol.sizeof == 200);
+extern (D):
+unittest
+{
+    assert(NominationProtocol.sizeof == getCPPSizeof!NominationProtocol());
+}

--- a/source/scpd/scp/SCP.d
+++ b/source/scpd/scp/SCP.d
@@ -131,4 +131,8 @@ nothrow:
     vector!SCPEnvelope getExternalizingState(uint64_t slotIndex);
 }
 
-static assert(SCP.sizeof == 48);
+extern (D):
+unittest
+{
+    assert(SCP.sizeof == getCPPSizeof!SCP());
+}

--- a/source/scpd/scp/Slot.d
+++ b/source/scpd/scp/Slot.d
@@ -120,4 +120,8 @@ private:
     vector!SCPEnvelope getEntireCurrentState();
 }
 
-static assert(Slot.sizeof == 400);
+extern (D):
+unittest
+{
+    assert(Slot.sizeof == getCPPSizeof!Slot());
+}

--- a/source/scpp/build.d
+++ b/source/scpp/build.d
@@ -59,7 +59,7 @@ version (Posix)
         "-D_GLIBCXX_USE_CXX11_ABI=0",
         "-std=c++14",
     ];
-    immutable CppCmd = [ "g++" ] ~ CppFlags;
+    immutable CppCmd = [ "clang++" ] ~ CppFlags;
 }
 else version (Windows)
 {

--- a/source/scpp/extra/DUtils.cpp
+++ b/source/scpp/extra/DUtils.cpp
@@ -4,7 +4,15 @@
 #include "DUtils.h"
 #include "xdrpp/marshal.h"
 #include "xdr/Stellar-SCP.h"
+#include "quorum/QuorumTracker.h"
+#include "quorum/QuorumIntersectionChecker.h"
+#include "scp/Slot.h"
+#include "scp/NominationProtocol.h"
+#include "scp/BallotProtocol.h"
+#include "scp/SCP.h"
 #include <functional>
+#include <unordered_map>
+#include <set>
 
 using namespace xdr;
 using namespace stellar;
@@ -65,12 +73,14 @@ PUSHBACKINST3(SCPQuorumSet, std::vector)
 
 
 #define CPPSETFOREACHINST(T) template int cpp_set_foreach<T>(void*, void*, void*);
+CPPSETFOREACHINST(int)
 CPPSETFOREACHINST(Value)
 CPPSETFOREACHINST(SCPBallot)
 CPPSETFOREACHINST(PublicKey)
 CPPSETFOREACHINST(unsigned int)
 
 #define CPPSETEMPTYINST(T) template bool cpp_set_empty<T>(const void*);
+CPPSETEMPTYINST(int)
 CPPSETEMPTYINST(Value)
 CPPSETEMPTYINST(SCPBallot)
 CPPSETEMPTYINST(PublicKey)
@@ -84,13 +94,6 @@ CPPUNORDEREDMAPASSIGNINST(int, int)
 CPPUNORDEREDMAPLENGTHINST(NodeID, std::shared_ptr<SCPQuorumSet>)
 CPPUNORDEREDMAPLENGTHINST(int, int)
 
-// @bug with substitution
-// https://issues.dlang.org/show_bug.cgi?id=20679
-// #define CPPUNORDEREDMAPCREATEINST(K, V) template std::unordered_map<K, V>* cpp_unordered_map_create<K, V>();
-#define CPPUNORDEREDMAPCREATEINST(K, V) template void* cpp_unordered_map_create<K, V>();
-CPPUNORDEREDMAPCREATEINST(NodeID, std::shared_ptr<SCPQuorumSet>)
-CPPUNORDEREDMAPCREATEINST(int, int)
-
 void callCPPDelegate (void* cb)
 {
     auto callback = (std::function<void()>*)cb;
@@ -103,3 +106,51 @@ std::shared_ptr<SCPQuorumSet> makeSharedSCPQuorumSet (
 {
     return std::make_shared<SCPQuorumSet>(quorum);
 }
+
+#define CPPDEFAULTCTORINST(T) template void defaultCtorCPPObject<T>(T*);
+#define CPPDTORINST(T) template void dtorCPPObject<T>(T*);
+#define CPPSIZEOFINST(T) template int getCPPSizeof<T>();
+#define CPPASSIGNINST(T) template void opAssignCPPObject<T> (T* lhs, T* rhs);
+#define CPPCOPYCTORINST(T) template void copyCtorCPPObject<T> (T* ptr, T* rhs);
+CPPSIZEOFINST(QuorumTracker);
+CPPSIZEOFINST(Slot);
+CPPSIZEOFINST(NominationProtocol);
+CPPSIZEOFINST(SCP);
+CPPSIZEOFINST(BallotProtocol);
+
+#define CPPOBJECTINST(T) CPPDEFAULTCTORINST(T) \
+                         CPPDTORINST(T)        \
+                         CPPSIZEOFINST(T)      \
+                         CPPASSIGNINST(T)      \
+                         CPPCOPYCTORINST(T)
+CPPOBJECTINST(std::shared_ptr<int>);
+CPPOBJECTINST(std::shared_ptr<SCPQuorumSet>);
+CPPOBJECTINST(std::shared_ptr<SCPEnvelope>);
+CPPOBJECTINST(std::shared_ptr<Slot>);
+CPPOBJECTINST(std::shared_ptr<LocalNode>);
+CPPOBJECTINST(std::shared_ptr<QuorumIntersectionChecker*>);
+
+CPPOBJECTINST(std::set<int>);
+CPPOBJECTINST(std::set<Value>);
+CPPOBJECTINST(std::set<PublicKey>);
+CPPOBJECTINST(std::set<SCPBallot>);
+CPPOBJECTINST(std::set<unsigned int>);
+
+#define CPPUNIQUEPTRINST(T) CPPDEFAULTCTORINST(std::unique_ptr<T>) \
+                            CPPDTORINST(std::unique_ptr<T>)        \
+                            CPPSIZEOFINST(std::unique_ptr<T>)
+CPPUNIQUEPTRINST(SCPEnvelope);
+CPPUNIQUEPTRINST(SCPBallot);
+CPPUNIQUEPTRINST(Value);
+
+
+#define CPPUNORDEREDMAPINST(K, V, id)   typedef std::unordered_map<K, V> ump_type_##id;  \
+                                        CPPOBJECTINST(ump_type_##id);
+CPPUNORDEREDMAPINST(NodeID, std::shared_ptr<SCPQuorumSet>, 0)
+CPPUNORDEREDMAPINST(int, int, 1)
+
+#define CPPMAPINST(K, V, id)   typedef std::map<K, V> map_type_##id;  \
+                                CPPOBJECTINST(map_type_##id);
+CPPMAPINST(int, int, 0)
+CPPMAPINST(PublicKey, SCPEnvelope, 1)
+CPPMAPINST(uint64_t, std::shared_ptr<Slot>, 2)

--- a/source/scpp/extra/DUtils.h
+++ b/source/scpp/extra/DUtils.h
@@ -74,3 +74,33 @@ VectorT duplicate(const VectorT& this_)
     VectorT dup = VectorT(this_);
     return dup;
 }
+
+template<typename T>
+int getCPPSizeof ()
+{
+    return sizeof(T);
+}
+
+template<typename T>
+void defaultCtorCPPObject (T* ptr)
+{
+    new(ptr) T();
+}
+
+template<typename T>
+void dtorCPPObject (T* ptr)
+{
+    ptr->~T();
+}
+
+template<typename T>
+void opAssignCPPObject (T* lhs, T* rhs)
+{
+    (*lhs) = *(rhs);
+}
+
+template<typename T>
+void copyCtorCPPObject (T* ptr, T* rhs)
+{
+    new(ptr) T(*(rhs));
+}

--- a/source/scpp/src/quorum/QuorumIntersectionCheckerImpl.cpp
+++ b/source/scpp/src/quorum/QuorumIntersectionCheckerImpl.cpp
@@ -760,7 +760,6 @@ namespace stellar
 std::shared_ptr<QuorumIntersectionChecker>
 QuorumIntersectionChecker::create(QuorumTracker::QuorumMap const& qmap)
 {
-    // note: cast due to current unordered_set binding in Cpp.d
-    return std::make_shared<QuorumIntersectionCheckerImpl>(**(QuorumTracker::QuorumMap**)&qmap);
+    return std::make_shared<QuorumIntersectionCheckerImpl>(qmap);
 }
 }

